runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/businessmetrics/BusinessMetricsUtils.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.businessmetrics
+
+import aws.smithy.kotlin.runtime.InternalApi
+import aws.smithy.kotlin.runtime.collections.AttributeKey
+import aws.smithy.kotlin.runtime.collections.get
+import aws.smithy.kotlin.runtime.operation.ExecutionContext
+
+/**
+ * Keeps track of all business metrics along an operations execution
+ */
+@InternalApi
+public val businessMetrics: AttributeKey<MutableSet<String>> = AttributeKey("aws.sdk.kotlin#BusinessMetrics")
+
+/**
+ * The account ID in an account ID based endpoint
+ */
+@InternalApi
+public val accountIdBasedEndPointAccountId: AttributeKey<String> = AttributeKey("aws.smithy.kotlin#AccountIdBasedEndpointAccountId")
+
+/**
+ * If an endpoint is service endpoint override based
+ */
+@InternalApi
+public val serviceEndpointOverride: AttributeKey<Boolean> = AttributeKey("aws.smithy.kotlin#ServiceEndpointOverride")
+
+/**
+ * Emit a business metric to the execution context attributes
+ */
+@InternalApi
+public fun ExecutionContext.emitBusinessMetric(metric: BusinessMetrics) {
+    if (this.attributes.contains(businessMetrics)) {
+        this.attributes[businessMetrics].add(metric.identifier)
+    } else {
+        this.attributes[businessMetrics] = mutableSetOf(metric.identifier)
+    }
+}
+
+/**
+ * Removes a business metric from the execution context attributes
+ */
+@InternalApi
+public fun ExecutionContext.removeBusinessMetric(metric: BusinessMetrics) {
+    if (this.attributes.contains(businessMetrics)) {
+        this.attributes[businessMetrics].remove(metric.identifier)
+    }
+}
+
+/**
+ * Checks if a business metric exists in the execution context attributes
+ */
+@InternalApi
+public fun ExecutionContext.containsBusinessMetric(metric: BusinessMetrics): Boolean =
+    (this.attributes.contains(businessMetrics)) && this.attributes[businessMetrics].contains(metric.identifier)
+
+/**
+ * All the valid business metrics along with their identifiers
+ */
+@InternalApi
+public enum class BusinessMetrics(public val identifier: String) {
```

Comment:
Some of these metrics are AWS-specific and don't belong in smithy-kotlin. We spent a lot of time making sure S3 Express was not mentioned in smithy-kotlin. 

Is there any way the AWS-specific enums can be relocated to aws-sdk-kotlin?

Reply:
I tried that too but it seems like that would require smithy kotlin have a dependency on the sdk. I think we want to avoid that

Reply:
Yeah, smithy-kotlin can not depend on aws-sdk-kotlin at all, that's a circular dependency.

Fixing this will certainly involve some rearchitecture. You will probably need to make an `AwsBusinessMetrics` class which implements `BusinessMetrics` so it can be used interchangeably in all the functions / utilities you wrote. That would also probably mean moving away from `enum class`, we can discuss offline